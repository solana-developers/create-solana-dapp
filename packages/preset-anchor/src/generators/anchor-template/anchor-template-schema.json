{
  "$schema": "http://json-schema.org/schema",
  "$id": "AnchorTemplateSchema",
  "title": "",
  "type": "object",
  "properties": {
    "name": {
      "type": "string",
      "description": "",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What name would you like to use?"
    },
    "directory": {
      "type": "string",
      "description": "",
      "$default": {
        "$source": "argv",
        "index": 1
      },
      "x-prompt": "What is the directory directory?"
    },
    "template": {
      "type": "string",
      "enum": ["base", "counter", "hello-world"],
      "description": "The template to use",
      "x-prompt": "Which template would you like to use?"
    }
  },
  "required": ["name", "directory", "template"]
}
