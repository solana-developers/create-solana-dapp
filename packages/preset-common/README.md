# @solana-developers/preset-common

This library is a preset for [create-solana-dapp](https://npm.im/create-solana-dapp) with shared code for the other
presets.

## Supported generators

### common-template

```bash
>  NX   generate @solana-developers/preset-common:common-template [name] [options,...]

From:  @solana-developers/preset-common (v0.0.1)
Name:  common-template

  Generate Common templates

Options:
    --directory                                                        [string]
    --name                                                             [string]
    --npmScope         The npm scope to use                            [string]
    --template         The template to use                   [string] [choices:
                                                           "license", "readme"]
    --webName          Name of the web application                     [string]
                       (overrides name)
    --anchor           Anchor template to use        [string] [choices: "none",
                                                      "counter", "hello-world"]
                                                              [default: "none"]
    --anchorName       Anchor project name         [string] [default: "anchor"]
    --licenseAuthor    The author to use in the                        [string]
                       license
    --preset           The preset to use             [string] [choices: "next",
                                                                       "react"]
```

## More information

- [GitHub repository](https://github.com/solana-developers/create-solana-dapp)
- [npm package](https://npm.im/create-solana-dapp)
- [StackOverflow](https://solana.stackexchange.com/questions/tagged/create-solana-dapp)
