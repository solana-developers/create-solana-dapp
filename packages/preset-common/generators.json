{
  "generators": {
    "common-template": {
      "factory": "./src/generators/common-template/common-template-generator",
      "schema": "./src/generators/common-template/common-template-schema.json",
      "description": "common-template generator"
    }
  }
}
