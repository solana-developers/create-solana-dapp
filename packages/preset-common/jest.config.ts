/* eslint-disable */
export default {
  displayName: 'preset-common',
  preset: '../../jest.preset.js',
  transform: {
    '^.+\\.[tj]s$': ['ts-jest', { tsconfig: '<rootDir>/tsconfig.spec.json' }],
  },
  moduleFileExtensions: ['ts', 'js', 'html'],
  coverageDirectory: '../../coverage/packages/preset-common',
  testEnvironment: 'node',
}
