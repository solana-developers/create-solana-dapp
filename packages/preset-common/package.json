{
  "name": "@solana-developers/preset-common",
  "version": "0.0.1",
  "description": "This library is a preset for create-solana-dapp with shared code for the other presets.",
  "dependencies": {
    "@nx/devkit": "17.1.3",
    "tslib": "^2.3.0"
  },
  "type": "commonjs",
  "main": "./src/index.js",
  "typings": "./src/index.d.ts",
  "generators": "./generators.json",
  "keywords": [
    "solana",
    "next",
    "react",
    "vue",
    "web3",
    "yarn",
    "blockchain",
    "nft"
  ],
  "license": "MIT",
  "homepage": "https://github.com/solana-developers/create-solana-dapp#readme",
  "bugs": {
    "url": "https://github.com/solana-developers/create-solana-dapp/issues"
  },
  "repository": {
    "name": "solana-developers/create-solana-dapp",
    "type": "git",
    "url": "https://github.com/solana-developers/create-solana-dapp"
  },
  "contributors": [
    {
      "name": "Joe Caulfield",
      "url": "https://github.com/jpcaulfi"
    },
    {
      "name": "Jacob Creech",
      "url": "https://github.com/jacobcreech"
    },
    {
      "name": "Nick Frostbutter",
      "url": "https://github.com/nickfrosty"
    },
    {
      "name": "Bram Borggreve",
      "url": "https://github.com/beeman"
    }
  ]
}
