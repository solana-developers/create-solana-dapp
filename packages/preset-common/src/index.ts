export * from './generators/common-template/common-template-generator'
export * from './generators/common-template/common-template-schema'
export * from './utils/generic-substitutions'
export * from './utils/application-cleanup'
export * from './utils/get-recursive-file-contents'
export * from './utils/package-versions'
export * from './utils/run-command'
export * from './utils/update-ignore-entires'
