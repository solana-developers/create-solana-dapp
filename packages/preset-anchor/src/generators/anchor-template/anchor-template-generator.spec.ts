import { Tree } from '@nx/devkit'
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing'
import { getRecursiveFileContents } from '@solana-developers/preset-common'
import anchorTemplateGenerator from './anchor-template-generator'
import { AnchorTemplateSchema } from './anchor-template-schema'

describe('anchor-template generator', () => {
  let tree: Tree
  const options: AnchorTemplateSchema = { name: 'test', directory: 'target', template: 'base' }

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace()
  })
  it.each(['base', 'counter', 'hello-world'])(
    'should generate files for %s template',
    async (template: AnchorTemplateSchema['template']) => {
      await anchorTemplateGenerator(tree, { ...options, template: template as AnchorTemplateSchema['template'] })

      const contents = getRecursiveFileContents(tree, options.directory)
      expect(contents).toMatchSnapshot()
    },
  )
})
