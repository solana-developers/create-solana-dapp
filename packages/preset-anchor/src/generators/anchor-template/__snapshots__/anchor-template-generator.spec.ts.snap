// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`anchor-template generator should generate files for base template 1`] = `
{
  "Anchor.toml": {
    "content": [
      "[features]",
      "seeds = false",
      "skip-lint = false",
      "[programs.localnet]",
      "base = "EuJfgknXicf6iZASeagt5U6XDDPpLTQrvWPW31vgAwo8"",
      "[registry]",
      "url = "https://api.apr.dev"",
      "[provider]",
      "cluster = "Localnet"",
      "wallet = "~/.config/solana/id.json"",
      "[scripts]",
      "test = "../node_modules/.bin/nx run test:jest"",
      "[test.validator]",
      "bind_address = "127.0.0.1"",
    ],
    "isBinary": false,
    "path": "target/Anchor.toml",
  },
  "Cargo.lock": {
    "content": null,
    "isBinary": true,
    "path": "target/Cargo.lock",
  },
  "Cargo.toml": {
    "content": [
      "[workspace]",
      "members = [",
      ""programs/*"",
      "]",
      "resolver = "2"",
      "[profile.release]",
      "overflow-checks = true",
      "lto = "fat"",
      "codegen-units = 1",
      "[profile.release.build-override]",
      "opt-level = 3",
      "incremental = false",
      "codegen-units = 1",
    ],
    "isBinary": false,
    "path": "target/Cargo.toml",
  },
  "migrations": {
    "children": {
      "deploy.ts": {
        "content": [
          "// Migrations are an early feature. Currently, they're nothing more than this",
          "// single deploy script that's invoked from the CLI, injecting a provider",
          "// configured from the workspace's Anchor.toml.",
          "const anchor = require("@coral-xyz/anchor");",
          "module.exports = async function (provider) {",
          "// Configure client to use the provider.",
          "anchor.setProvider(provider);",
          "// Add your deploy script here.",
          "};",
        ],
        "isBinary": false,
        "path": "target/migrations/deploy.ts",
      },
    },
    "path": "target/migrations",
  },
}
`;

exports[`anchor-template generator should generate files for counter template 1`] = `
{
  "Anchor.toml": {
    "content": [
      "[features]",
      "seeds = false",
      "skip-lint = false",
      "[programs.localnet]",
      "counter = "EuJfgknXicf6iZASeagt5U6XDDPpLTQrvWPW31vgAwo8"",
      "[registry]",
      "url = "https://api.apr.dev"",
      "[provider]",
      "cluster = "Localnet"",
      "wallet = "~/.config/solana/id.json"",
      "[scripts]",
      "test = "../node_modules/.bin/nx run test:jest"",
      "[test.validator]",
      "bind_address = "127.0.0.1"",
    ],
    "isBinary": false,
    "path": "target/Anchor.toml",
  },
  "programs": {
    "children": {
      "counter": {
        "children": {
          "Cargo.toml": {
            "content": [
              "[package]",
              "name = "counter"",
              "version = "0.1.0"",
              "description = "Created with Anchor"",
              "edition = "2021"",
              "[lib]",
              "crate-type = ["cdylib", "lib"]",
              "name = "counter"",
              "[features]",
              "no-entrypoint = []",
              "no-idl = []",
              "no-log-ix-name = []",
              "cpi = ["no-entrypoint"]",
              "default = []",
              "[dependencies]",
              "anchor-lang = "0.28.0"",
            ],
            "isBinary": false,
            "path": "target/programs/counter/Cargo.toml",
          },
          "Xargo.toml": {
            "content": [
              "[target.bpfel-unknown-unknown.dependencies.std]",
              "features = []",
            ],
            "isBinary": false,
            "path": "target/programs/counter/Xargo.toml",
          },
          "src": {
            "children": {
              "lib.rs": {
                "content": [
                  "#![allow(clippy::result_large_err)]",
                  "use anchor_lang::prelude::*;",
                  "declare_id!("EuJfgknXicf6iZASeagt5U6XDDPpLTQrvWPW31vgAwo8");",
                  "#[program]",
                  "pub mod counter {",
                  "use super::*;",
                  "pub fn initialize_counter(_ctx: Context<InitializeCounter>) -> Result<()> {",
                  "Ok(())",
                  "}",
                  "pub fn increment(ctx: Context<Increment>) -> Result<()> {",
                  "ctx.accounts.counter.count = ctx.accounts.counter.count.checked_add(1).unwrap();",
                  "Ok(())",
                  "}",
                  "}",
                  "#[derive(Accounts)]",
                  "pub struct InitializeCounter<'info> {",
                  "#[account(mut)]",
                  "pub payer: Signer<'info>,",
                  "#[account(",
                  "init,",
                  "space = 8 + Counter::INIT_SPACE,",
                  "payer = payer",
                  ")]",
                  "pub counter: Account<'info, Counter>,",
                  "pub system_program: Program<'info, System>,",
                  "}",
                  "#[derive(Accounts)]",
                  "pub struct Increment<'info> {",
                  "#[account(mut)]",
                  "pub counter: Account<'info, Counter>,",
                  "}",
                  "#[account]",
                  "#[derive(InitSpace)]",
                  "pub struct Counter {",
                  "count: u64,",
                  "}",
                ],
                "isBinary": false,
                "path": "target/programs/counter/src/lib.rs",
              },
            },
            "path": "target/programs/counter/src",
          },
        },
        "path": "target/programs/counter",
      },
    },
    "path": "target/programs",
  },
  "src": {
    "children": {
      "index.ts": {
        "content": [
          "import { Program } from '@coral-xyz/anchor';",
          "import { Counter, IDL as CounterIDL } from '../target/types/counter';",
          "export { Counter, CounterIDL };",
          "export type CounterProgram = Program<Counter>;",
        ],
        "isBinary": false,
        "path": "target/src/index.ts",
      },
    },
    "path": "target/src",
  },
  "target": {
    "children": {
      "types": {
        "children": {
          "counter.ts": {
            "content": [
              "export type Counter = {",
              ""version": "0.1.0",",
              ""name": "counter",",
              ""instructions": [",
              "{",
              ""name": "initializeCounter",",
              ""accounts": [",
              "{",
              ""name": "payer",",
              ""isMut": true,",
              ""isSigner": true",
              "},",
              "{",
              ""name": "counter",",
              ""isMut": true,",
              ""isSigner": true",
              "},",
              "{",
              ""name": "systemProgram",",
              ""isMut": false,",
              ""isSigner": false",
              "}",
              "],",
              ""args": []",
              "},",
              "{",
              ""name": "increment",",
              ""accounts": [",
              "{",
              ""name": "counter",",
              ""isMut": true,",
              ""isSigner": false",
              "}",
              "],",
              ""args": []",
              "}",
              "],",
              ""accounts": [",
              "{",
              ""name": "counter",",
              ""type": {",
              ""kind": "struct",",
              ""fields": [",
              "{",
              ""name": "count",",
              ""type": "u64"",
              "}",
              "]",
              "}",
              "}",
              "]",
              "};",
              "export const IDL: Counter = {",
              ""version": "0.1.0",",
              ""name": "counter",",
              ""instructions": [",
              "{",
              ""name": "initializeCounter",",
              ""accounts": [",
              "{",
              ""name": "payer",",
              ""isMut": true,",
              ""isSigner": true",
              "},",
              "{",
              ""name": "counter",",
              ""isMut": true,",
              ""isSigner": true",
              "},",
              "{",
              ""name": "systemProgram",",
              ""isMut": false,",
              ""isSigner": false",
              "}",
              "],",
              ""args": []",
              "},",
              "{",
              ""name": "increment",",
              ""accounts": [",
              "{",
              ""name": "counter",",
              ""isMut": true,",
              ""isSigner": false",
              "}",
              "],",
              ""args": []",
              "}",
              "],",
              ""accounts": [",
              "{",
              ""name": "counter",",
              ""type": {",
              ""kind": "struct",",
              ""fields": [",
              "{",
              ""name": "count",",
              ""type": "u64"",
              "}",
              "]",
              "}",
              "}",
              "]",
              "};",
            ],
            "isBinary": false,
            "path": "target/target/types/counter.ts",
          },
        },
        "path": "target/target/types",
      },
    },
    "path": "target/target",
  },
  "tests": {
    "children": {
      "counter.spec.ts": {
        "content": [
          "import * as anchor from "@coral-xyz/anchor";",
          "import { Program } from "@coral-xyz/anchor";",
          "import { Keypair } from '@solana/web3.js';",
          "import { Counter } from "../target/types/counter";",
          "describe("counter", () => {",
          "// Configure the client to use the local cluster.",
          "const provider = anchor.AnchorProvider.env();",
          "anchor.setProvider(provider);",
          "const payer = provider.wallet as anchor.Wallet;",
          "const program = anchor.workspace.Counter as Program<Counter>;",
          "// Generate a new keypair for the counter account",
          "const counterKeypair = new Keypair();",
          "it('Initialize Counter', async () => {",
          "await program.methods",
          ".initializeCounter()",
          ".accounts({",
          "counter: counterKeypair.publicKey,",
          "payer: payer.publicKey,",
          "})",
          ".signers([counterKeypair])",
          ".rpc();",
          "const currentCount = await program.account.counter.fetch(",
          "counterKeypair.publicKey",
          ");",
          "expect(currentCount.count.toNumber()).toEqual(0);",
          "});",
          "it('Increment Counter', async () => {",
          "await program.methods",
          ".increment()",
          ".accounts({ counter: counterKeypair.publicKey })",
          ".rpc();",
          "const currentCount = await program.account.counter.fetch(",
          "counterKeypair.publicKey",
          ");",
          "expect(currentCount.count.toNumber()).toEqual(1);",
          "});",
          "it('Increment Counter Again', async () => {",
          "await program.methods",
          ".increment()",
          ".accounts({ counter: counterKeypair.publicKey })",
          ".rpc();",
          "const currentCount = await program.account.counter.fetch(",
          "counterKeypair.publicKey",
          ");",
          "expect(currentCount.count.toNumber()).toEqual(2);",
          "});",
          "});",
        ],
        "isBinary": false,
        "path": "target/tests/counter.spec.ts",
      },
    },
    "path": "target/tests",
  },
}
`;

exports[`anchor-template generator should generate files for hello-world template 1`] = `
{
  "Anchor.toml": {
    "content": [
      "[features]",
      "seeds = false",
      "skip-lint = false",
      "[programs.localnet]",
      "hello_world = "EuJfgknXicf6iZASeagt5U6XDDPpLTQrvWPW31vgAwo8"",
      "[registry]",
      "url = "https://api.apr.dev"",
      "[provider]",
      "cluster = "Localnet"",
      "wallet = "~/.config/solana/id.json"",
      "[scripts]",
      "test = "../node_modules/.bin/nx run test:jest"",
      "[test.validator]",
      "bind_address = "127.0.0.1"",
    ],
    "isBinary": false,
    "path": "target/Anchor.toml",
  },
  "programs": {
    "children": {
      "hello-world": {
        "children": {
          "Cargo.toml": {
            "content": [
              "[package]",
              "name = "hello-world"",
              "version = "0.1.0"",
              "description = "Created with Anchor"",
              "edition = "2021"",
              "[lib]",
              "crate-type = ["cdylib", "lib"]",
              "name = "hello_world"",
              "[features]",
              "no-entrypoint = []",
              "no-idl = []",
              "no-log-ix-name = []",
              "cpi = ["no-entrypoint"]",
              "default = []",
              "[dependencies]",
              "anchor-lang = "0.28.0"",
            ],
            "isBinary": false,
            "path": "target/programs/hello-world/Cargo.toml",
          },
          "Xargo.toml": {
            "content": [
              "[target.bpfel-unknown-unknown.dependencies.std]",
              "features = []",
            ],
            "isBinary": false,
            "path": "target/programs/hello-world/Xargo.toml",
          },
          "src": {
            "children": {
              "lib.rs": {
                "content": [
                  "use anchor_lang::prelude::*;",
                  "declare_id!("EuJfgknXicf6iZASeagt5U6XDDPpLTQrvWPW31vgAwo8");",
                  "#[program]",
                  "pub mod hello_world {",
                  "use super::*;",
                  "pub fn initialize(_ctx: Context<Initialize>) -> Result<()> {",
                  "msg!("Hello World!");",
                  "Ok(())",
                  "}",
                  "}",
                  "#[derive(Accounts)]",
                  "pub struct Initialize {}",
                ],
                "isBinary": false,
                "path": "target/programs/hello-world/src/lib.rs",
              },
            },
            "path": "target/programs/hello-world/src",
          },
        },
        "path": "target/programs/hello-world",
      },
    },
    "path": "target/programs",
  },
  "src": {
    "children": {
      "index.ts": {
        "content": [
          "export * from '../target/types/hello_world';",
        ],
        "isBinary": false,
        "path": "target/src/index.ts",
      },
    },
    "path": "target/src",
  },
  "target": {
    "children": {
      "types": {
        "children": {
          "hello_world.ts": {
            "content": [
              "export type HelloWorld = {",
              ""version": "0.1.0",",
              ""name": "hello_world",",
              ""instructions": [",
              "{",
              ""name": "initialize",",
              ""accounts": [],",
              ""args": []",
              "}",
              "]",
              "};",
              "export const IDL: HelloWorld = {",
              ""version": "0.1.0",",
              ""name": "hello_world",",
              ""instructions": [",
              "{",
              ""name": "initialize",",
              ""accounts": [],",
              ""args": []",
              "}",
              "]",
              "};",
            ],
            "isBinary": false,
            "path": "target/target/types/hello_world.ts",
          },
        },
        "path": "target/target/types",
      },
    },
    "path": "target/target",
  },
  "tests": {
    "children": {
      "hello-world.spec.ts": {
        "content": [
          "import * as anchor from "@coral-xyz/anchor";",
          "import { Program } from "@coral-xyz/anchor";",
          "import { HelloWorld } from "../target/types/hello_world";",
          "describe("hello-world", () => {",
          "// Configure the client to use the local cluster.",
          "anchor.setProvider(anchor.AnchorProvider.env());",
          "const program = anchor.workspace.HelloWorld as Program<HelloWorld>;",
          "it("Is initialized!", async () => {",
          "// Add your test here.",
          "const tx = await program.methods.initialize().rpc();",
          "console.log("Your transaction signature", tx);",
          "});",
          "});",
        ],
        "isBinary": false,
        "path": "target/tests/hello-world.spec.ts",
      },
    },
    "path": "target/tests",
  },
}
`;
